# Project: No-Show Appointments

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
# Introduction

> This dataset collects information from 100k medical appointments in Brazil and is focused on the question of whether or not patients show up for their appointment. A number of characteristics about the patient are included in each row.

<a id='wrangling'></a>
## Data Wrangling

#### Step 1: Import data analysis packages

In [32]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
% matplotlib inline

#### Step 2: Import data to Jupyter notebook

In [33]:
df = pd.read_csv('noshowappointments.csv')

#### Step 3: Check to see if there are any inconsistencies in the data set

In [34]:
df.head()

PatientId  AppointmentID Gender          ScheduledDay  \
0  2.987250e+13        5642903      F  2016-04-29T18:38:08Z   
1  5.589980e+14        5642503      M  2016-04-29T16:08:27Z   
2  4.262960e+12        5642549      F  2016-04-29T16:19:04Z   
3  8.679510e+11        5642828      F  2016-04-29T17:29:31Z   
4  8.841190e+12        5642494      F  2016-04-29T16:07:23Z   

         AppointmentDay  Age      Neighbourhood  Scholarship  Hipertension  \
0  2016-04-29T00:00:00Z   62    JARDIM DA PENHA            0             1   
1  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             0   
2  2016-04-29T00:00:00Z   62      MATA DA PRAIA            0             0   
3  2016-04-29T00:00:00Z    8  PONTAL DE CAMBURI            0             0   
4  2016-04-29T00:00:00Z   56    JARDIM DA PENHA            0             1   

   Diabetes  Alcoholism  Handcap  SMS_received No-show  
0         0           0        0             0      No  
1         0           0        0             0      No  
2         0           0        0             0      No  
3         0           0        0             0      No  
4         1           0        0             0      No

> #### Notes
>
> ##### Description of Columns
>____________________________________________________________________________
> Patient ID - Personal identification number that describes the patient in-house
>
> Appointment ID - The ID used to describe the appointment
>
> Gender - Describing whether the patient is male or female
>
> Scheduled Day - Day the patient scheduled the appointment
>
> Appointment Day - Day the patient is scheduled to arrive at the appointment
>
> Age - Indicates the age of the patient
>
> Neighbourhood - Indicates the location of the appointment
>
> Scholarship - Indicates whether or not the patient is enrolled in Brasilian welfare program Bolsa Família
>
> Hipertension - Indicates whether the patient has hipertension
>
> Diabetes - Indicates whether the patient has diabetes
>
> Alcoholism - Indicates whether the patient is an alcoholic or has been an alcoholic
>
> Handcap - Indicates whether the patient is handicap
>
> SMS Recieved - Indicates whether the patient recieved an SMS reminder
>
> No-Show - Indicates whether the patient showed up for the appointment (no meaning showed up, yes meaning no show)

In [35]:
df.describe()

PatientId  AppointmentID            Age    Scholarship  \
count  1.105270e+05   1.105270e+05  110527.000000  110527.000000   
mean   1.474963e+14   5.675305e+06      37.088874       0.098266   
std    2.560949e+14   7.129575e+04      23.110205       0.297675   
min    3.920000e+04   5.030230e+06      -1.000000       0.000000   
25%    4.172615e+12   5.640286e+06      18.000000       0.000000   
50%    3.173180e+13   5.680573e+06      37.000000       0.000000   
75%    9.439170e+13   5.725524e+06      55.000000       0.000000   
max    9.999820e+14   5.790484e+06     115.000000       1.000000   

        Hipertension       Diabetes     Alcoholism        Handcap  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean        0.197246       0.071865       0.030400       0.022248   
std         0.397921       0.258265       0.171686       0.161543   
min         0.000000       0.000000       0.000000       0.000000   
25%         0.000000       0.000000       0.000000       0.000000   
50%         0.000000       0.000000       0.000000       0.000000   
75%         0.000000       0.000000       0.000000       0.000000   
max         1.000000       1.000000       1.000000       4.000000   

        SMS_received  
count  110527.000000  
mean        0.321026  
std         0.466873  
min         0.000000  
25%         0.000000  
50%         0.000000  
75%         1.000000  
max         1.000000

In [36]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 14 columns):
PatientId         110527 non-null float64
AppointmentID     110527 non-null int64
Gender            110527 non-null object
ScheduledDay      110527 non-null object
AppointmentDay    110527 non-null object
Age               110527 non-null int64
Neighbourhood     110527 non-null object
Scholarship       110527 non-null int64
Hipertension      110527 non-null int64
Diabetes          110527 non-null int64
Alcoholism        110527 non-null int64
Handcap           110527 non-null int64
SMS_received      110527 non-null int64
No-show           110527 non-null object
dtypes: float64(1), int64(8), object(5)
memory usage: 11.8+ MB


In [37]:
df.shape

(110527, 14)

Observation / Needed Changes: There are 14 columns and 110527 rows

In [38]:
print(sum(df.duplicated()))

0


No null values in the data set

> #### Changes Needed:
    - Change column names for ease of use
    - Remove Patient ID and Appoinment ID because they are unnecessary
    - Remove any age that is smaller than zero
    - Reformatting scheduled_day and appointment_day to datetime
    - Reformatting no_show data to integar

<a id='wrangling'></a>
## Data Cleaning

#### Step 1: Change column names for ease of use

In [39]:
new_labels = ['patientID', 'appointmentID', 'gender', 'schedule_day', 'appointment_day', 'age', 'neighborhood', 'scholarship', 'hypertension', 'diabetes', 'alcoholism', 'handicap','sms_received', 'no_show']
df.columns = new_labels
df.tail()

patientID  appointmentID gender          schedule_day  \
110522  2.572130e+12        5651768      F  2016-05-03T09:15:35Z   
110523  3.596270e+12        5650093      F  2016-05-03T07:27:33Z   
110524  1.557660e+13        5630692      F  2016-04-27T16:03:52Z   
110525  9.213490e+13        5630323      F  2016-04-27T15:09:23Z   
110526  3.775120e+14        5629448      F  2016-04-27T13:30:56Z   

             appointment_day  age neighborhood  scholarship  hypertension  \
110522  2016-06-07T00:00:00Z   56  MARIA ORTIZ            0             0   
110523  2016-06-07T00:00:00Z   51  MARIA ORTIZ            0             0   
110524  2016-06-07T00:00:00Z   21  MARIA ORTIZ            0             0   
110525  2016-06-07T00:00:00Z   38  MARIA ORTIZ            0             0   
110526  2016-06-07T00:00:00Z   54  MARIA ORTIZ            0             0   

        diabetes  alcoholism  handicap  sms_received no_show  
110522         0           0         0             1      No  
110523         0           0         0             1      No  
110524         0           0         0             1      No  
110525         0           0         0             1      No  
110526         0           0         0             1      No

#### Step 2: Remove Patient ID and Appoinment ID

In [40]:
df.drop(['patientID', 'appointmentID'], axis=1, inplace=True)

In [41]:
df.describe()

age    scholarship   hypertension       diabetes  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean       37.088874       0.098266       0.197246       0.071865   
std        23.110205       0.297675       0.397921       0.258265   
min        -1.000000       0.000000       0.000000       0.000000   
25%        18.000000       0.000000       0.000000       0.000000   
50%        37.000000       0.000000       0.000000       0.000000   
75%        55.000000       0.000000       0.000000       0.000000   
max       115.000000       1.000000       1.000000       1.000000   

          alcoholism       handicap   sms_received  
count  110527.000000  110527.000000  110527.000000  
mean        0.030400       0.022248       0.321026  
std         0.171686       0.161543       0.466873  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000  
50%         0.000000       0.000000       0.000000  
75%         0.000000       0.000000       1.000000  
max         1.000000       4.000000       1.000000

#### Step 3. Remove age that is smaller than zero

In [42]:
meanage = df['age'].mean()
df[df['age'] <= 0] = meanage

In [43]:
df.describe()

age    scholarship   hypertension       diabetes  \
count  110527.000000  110527.000000  110527.000000  110527.000000   
mean       38.276780       1.285691       1.385142       1.259761   
std        22.104561       6.519426       6.506669       6.522499   
min         1.000000       0.000000       0.000000       0.000000   
25%        20.000000       0.000000       0.000000       0.000000   
50%        37.088874       0.000000       0.000000       0.000000   
75%        55.000000       0.000000       0.000000       0.000000   
max       115.000000      37.088874      37.088874      37.088874   

          alcoholism       handicap   sms_received  
count  110527.000000  110527.000000  110527.000000  
mean        1.218296       1.210135       1.500806  
std         6.527195       6.528421       6.489885  
min         0.000000       0.000000       0.000000  
25%         0.000000       0.000000       0.000000  
50%         0.000000       0.000000       0.000000  
75%         0.000000       0.000000       1.000000  
max        37.088874      37.088874      37.088874

#### Step 4. Reformatting Scheduled Day and Appointment Day to datetime

In [44]:
df['schedule_day'] = pd.to_datetime(df['schedule_day'])
df['appointment_day'] = pd.to_datetime(df['appointment_day'])

In [45]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 110527 entries, 0 to 110526
Data columns (total 12 columns):
gender             110527 non-null object
schedule_day       110527 non-null datetime64[ns]
appointment_day    110527 non-null datetime64[ns]
age                110527 non-null float64
neighborhood       110527 non-null object
scholarship        110527 non-null float64
hypertension       110527 non-null float64
diabetes           110527 non-null float64
alcoholism         110527 non-null float64
handicap           110527 non-null float64
sms_received       110527 non-null float64
no_show            110527 non-null object
dtypes: datetime64[ns](2), float64(7), object(3)
memory usage: 10.1+ MB


#### Step 5. Reformatting the no_show column to integar

In [ ]:
df["no_show"] = np.where((df.no_show=="Yes"), 1, 0)
df.head()

> #### Changes Needed:
    - Create dataframe for people who showed and people who didn't show
    - Create dataframe for the length of time between schedule_day and appointment_day
    - 

<a id='wrangling'></a>
## Data Analysis

#### Step 1: Create Dataframes

    a. Create data frame for length of time between schedule_day and appointment_day

In [47]:
df['appointment_day'] = pd.to_datetime(df['appointment_day'])
df['schedule_day'] = pd.to_datetime(df['schedule_day'])
df['elapsed_days'] = df['appointment_day'] - df['schedule_day']

    b. Create dataframe for people who did not show for appointment

In [48]:
no_shows = df.query('no_show == "1"')
days = no_shows['elapsed_days'].mean()
timeframe_nos = days / np.timedelta64(1, 'D')

    c. Create dataframe for people who showed for appointment

In [49]:
shows = df.query('no_show == "0"')
days = shows['elapsed_days'].mean()
timeframe_shows = days / np.timedelta64(1, 'D')

#### Step 2: Visualize Shows vs No Shows

In [ ]:
new_df = df.groupby('no_show')['no_show'].agg(['count'])
new_df.reset_index(inplace=True)
new_df.columns = ['appt_status', 'total']
new_df.iloc[0,0] = "No Show"
new_df.iloc[1,0] = "Show"

nos_prop = new_df.iloc[0,1] / (new_df.iloc[0,1] + new_df.iloc[1,1])
shows_prop = new_df.iloc[1,1] / (new_df.iloc[0,1] + new_df.iloc[1,1])

new_df['proportion'] = [nos_prop, shows_prop]
new_df.head()

In [ ]:
labels = ['No Shows', 'Shows']
sizes = [80.4, 19.6]
explode = (0.1, 0.1)

fig, ax = plt.subplots()

ax.pie(sizes, explode=explode, labels=labels, autopct='%1.1f%%', shadow=True, startangle=80)
ax.axis('equal')
ax.set_title('Appointment No Shows vs Shows')


plt.show()

#### Step 3: Visualize Neighborhood Locations

In [ ]:
df['neighborhood'].value_counts().plot(kind='bar', figsize=(20,8))

In [ ]:
shows_by_hood = shows.groupby("neighborhood")["no_show"].count().reset_index(name="count").sort_values("count", ascending=False)
shows_by_hood.head(10)

ADD GRAPH HERE - No Shows in Area vs Total Count of Patients

#### Step 4: Visualize Correlation Between Age and Whether Patient Showed for Appointment

In [ ]:
age_df_noshows = no_shows['age'].mean()
age_df_noshows

In [ ]:
age_df_shows = shows['age'].mean()
age_df_shows

In [ ]:
age_df_noshows = no_shows['age']
age_df_shows = shows['age']

In [ ]:
plt.hist(age_df_shows, label='Shows')
plt.hist(age_df_noshows, label='No Shows')
plt.title('Age Distribution by Shows vs No Shows')
plt.xlabel('Age')
plt.ylabel('Number of Appointments')
plt.legend();

#### Step 5: Visualize Correlation Between Whether Patient Received SMS Message and Whether Patient Showed for Appointment

In [ ]:
noshow_sms_count = no_shows.query('sms_received == 1')
a = noshow_sms_count.shape[0]

In [ ]:
shows_sms_count = shows.query('sms_received == 1')
b = shows_sms_count.shape[0]

In [ ]:
locations = [1, 2]
heights = [a, b]
labels = ['Shows', 'No Shows']

bar1 = plt.bar(locations, heights, tick_label=labels)
plt.title('SMS Message Received')
plt.xlabel('Showed or No-Showed for Appointment')
plt.ylabel('SMS Received');


In [ ]:
noshow_sms_count = no_shows.query('sms_received == 0')
c = noshow_sms_count.shape[0]

shows_sms_count = shows.query('sms_received == 0')
d = shows_sms_count.shape[0]

In [ ]:
locations = [1, 2]
heights = [c, d]
labels = ['Shows', 'No Shows']

bar2 = plt.bar(locations, heights, tick_label=labels)
plt.title('SMS Messages Not Received')
plt.xlabel('Showed or No-Showed for Appointment')
plt.ylabel('SMS Not Received');

#### Step 5: Visualize Correlation Length of Time Between Scheduled Appointment and Date of Appointment and Whether Patient Showed for Appointment

In [ ]:
df['between_days'] = df['appointment_day'] - df['schedule_day']

In [ ]:
shows = df.query('no_show == "0"')
days = shows['between_days'].mean()
timeframe_shows = days / np.timedelta64(1, 'D')

In [ ]:
no_shows = df.query('no_show == "1"')
days = no_shows['between_days'].mean()
timeframe_noshows = days / np.timedelta64(1, 'D')

In [ ]:
locations = [1, 2]
heights = [timeframe_shows, timeframe_noshows]
labels = ['Shows', 'No Shows']

plt.bar(locations, heights, tick_label=labels)
plt.title('Average Appointment Show Rates')
plt.xlabel('Showed or No-Showed for Appointment')
plt.ylabel('Days Between Booking and Appointment');

#### Step 6: Investigating Coorelation Between Appointment Attendance and Gender

In [ ]:
line_df = df.groupby(["gender","no_show"])["no_show"].count().reset_index(name="appt_count")
line_df.head()

<a id='eda'></a>
## Exploratory Data Analysis

> **Tip**: Now that you've trimmed and cleaned your data, you're ready to move on to exploration. Compute statistics and create visualizations with the goal of addressing the research questions that you posed in the Introduction section. It is recommended that you be systematic with your approach. Look at one variable at a time, and then follow it up by looking at relationships between variables.

### Research Question 1 - Did the SMS Message Help Attendance Rate

### Research Question 2  - Did Age Affect Attendance Rate

In [ ]:
# Continue to explore the data to address your additional research
#   questions. Add more headers as needed if you have more questions to
#   investigate.


<a id='conclusions'></a>
## Conclusions

> **Tip**: Finally, summarize your findings and the results that have been performed. Make sure that you are clear with regards to the limitations of your exploration. If you haven't done any statistical tests, do not imply any statistical conclusions. And make sure you avoid implying causation from correlation!

> **Tip**: Once you are satisfied with your work here, check over your report to make sure that it is satisfies all the areas of the rubric (found on the project submission page at the end of the lesson). You should also probably remove all of the "Tips" like this one so that the presentation is as polished as possible.

## Submitting your Project 

> Before you submit your project, you need to create a .html or .pdf version of this notebook in the workspace here. To do that, run the code cell below. If it worked correctly, you should get a return code of 0, and you should see the generated .html file in the workspace directory (click on the orange Jupyter icon in the upper left).

> Alternatively, you can download this report as .html via the **File** > **Download as** submenu, and then manually upload it into the workspace directory by clicking on the orange Jupyter icon in the upper left, then using the Upload button.

> Once you've done this, you can submit your project by clicking on the "Submit Project" button in the lower right here. This will create and submit a zip file with this .ipynb doc and the .html or .pdf version you created. Congratulations!

In [ ]:
from subprocess import call
call(['python', '-m', 'nbconvert', 'Investigate_a_Dataset.ipynb'])